doc/developer/design/20240502_deprecate_subsources.md

```diff
@@ -0,0 +1,190 @@
+# sql+storage: deprecate concept of subsources
+
+- Associated:
+    - [storage/sources: make (sub)sources a flavor of tables #20208
+      ](https://github.com/MaterializeInc/materialize/issues/20208)
+    - [sql/storage: consider if TEXT COLUMNS, IGNORE COLUMNS belongs on
+      subsources
+      #26774](https://github.com/MaterializeInc/materialize/issues/26774)
+
+## The Problem
+
+There are two distinct notions of `SUBSOURCE`s in MZ, both of which can be
+folded neatly into other concepts.
+
+- Ingestion export subsources should be tables (as should the current primary
+  source's collection)
+- Progress subsources should become the relation returned when querying sources
+
+Doing both of these lets us remove the concept of subsources from Materialize
+entirely. It, by definition, also lets us get rid of the notion of "progress
+subsources," which is a concept that users often find confusion.
+
+## Success Criteria
+
+- Users no longer need to use the keyword `SUBSOURCE` to interact with
+  Materialize in any capacity; and we do this without deprecating any
+  functionality whatsoever.
+
+## Solution Proposal
+
+### SOURCEs should output to TABLES
+
+In the current version of Materialize, we have an impedance mismatch between
+single-output and multi-output sources.
+
+- Single-ouput sources write their data directly to the `SOURCE` itself, i.e.
+  users query the `SOURCE` object to read the data.
+- Multi-output sources write their data to `SUBSOURCES`. The primary source to
+  which they belong is an empty, unused (though still queryable) collection.
+
+I propose that the output of all sources, both single- and multi-output, should
+be to `TABLE`s (albeit special, read-only tables). Users already have an
+intuition around how `TABLE`s work, and understanding "read-only" tables is much
+simpler than doubling down on subsources––a concept we invented.
+
+#### Why not just rename `SUBSOURCE` to `TABLE`?
+
+This leaves the impedance mismatch in place where users have to `SELECT` from
+the source if they're using e.g. Kafka, but have to find the right `TABLE` to
+`SELECT` from for PG sources.
+
+### SOURCES should contain their progress information
+
+The other type of `SUBSOURCE` Materialize supports is known as the
+remap/progress collection. While most users don't access this data, it's an
+important debugging tool.
+
+I propose that the progress collection should be moved "into" the source, i.e.
+querying a source should return the what-is-today-the-progress-collection's
+data.
+
+### New API
+
+The API to support this is similar to the existing `CREATE SOURCE` API. However,
+`ALTER SOURCE...ADD SUBSOURCE` will disappear and be replaced with a flavor of
+`CREATE TABLE`.
```

Comment:
Sayyyyyy .. what if we punted on this for now? I feel like this migration is a whole lot less disruptive if we don't fundamentally change the syntax. We just delete the `SUBSOURCE` keywords from the commands, to force people to use the `TABLE` variants. E.g., instead of typing `ALTER SOURCE ... ADD SUBSOURCE`, you have to type `ALTER SOURCE ... ADD TABLE`.

If we think one day we'll want to support configuring formats and envelopes on a per subsource (er, table) basis, I feel like _that_'s a good moment to think about pivoting from `ALTER SOURCE ... ADD TABLE` to some sort of `CREATE TABLE FROM SOURCE` syntax:

```
CREATE TABLE
FROM SOURCE
[TRANSFORM USING ... | FORMAT ... | ENVELOPE ...]
```

But in the meantime, I think `CREATE SOURCE ... TABLES (...)` coupled with `ALTER SOURCE ... [ADD | DROP] TABLE` solves like 90%+ of the user confusion around subsources?

Reply:
Wait, no, what was I thinking?! Must have been the jetlag. If we don't do the `CREATE TABLE FROM SOURCE` piece then things don't integrate with dbt. Never mind, never mind.